src/lib/rank_utils.ts

```diff
@@ -22,9 +22,74 @@ interface IRankInfo {
     label: string;
 }
 
-export const MaxRank: number = 36;
+class Rating {
+    unset:boolean;
+    rating:number;
+    deviation:number;
+    volatility:number;
+    rank:number;
+    rank_label:string;
+    partial_rank:number;
+    partial_rank_label:string;
+    rank_deviation_labels:Array<string>;
+    rank_deviation:number;
+}
+
+export const MaxRank: number = 39;
+
+
+const MIN_RATING = 100;
+const MAX_RATING = 6000;
+
+export function rank_to_rating(rank:number) {
+    return 900 * Math.exp(0.032 * rank);
+}
+
+export function rating_to_rank(rating:number) {
+    return Math.log(Math.min(MAX_RATING, Math.max(MIN_RATING, rating)) / 900.0) / 0.032;
+}
+
+export function get_handicap_adjustment(rating:number, handicap:number):number {
+    return rank_to_rating(rating_to_rank(rating) + handicap) - rating;
```

Comment:
If your goal is to handicap the games so that the expected win rate for black is 50%, then you should use handicaps + reverse komi. IGS aka Panadanet does this to get half stone handicap increments, and their ranks are actually split into half stone increments, i.e. 2d and 2d+, then 3d and 3d+. See section 2 Handicaps: http://pandanet-igs.com/communities/pandanet/60

The disadvantage of that system is players have to always check what the komi is, but the advantage is you get finer control of the handicaps, and the expected win rates will be 50% like you are looking for.



Personally I don't like that because I always assume komi is 0.5 in a handicap game, but it sounds like this system would